### PR TITLE
and operator unexpected boolean cast test

### PR DESCRIPTION
**Pull request by bgalek**
we need to check it out!

### CODE REVIEW

**Review by pawel-piecyk — APPROVED** [review]

File: src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java [hunk 1]
```diff
@@ -40,20 +40,17 @@ static LogicalOperatorExpressionNode orOperator(OpelNode left, OpelNode right, I
...
 
     private <T> T wrappingExceptionsWithOpelException(Object left, Object right, Supplier<T> wrappedBody) {
         try {
             return wrappedBody.get();
         } catch (Exception e) {
-            throw new OpelException("Error on evaluating logical expression. " +
-                    "operator: " + logicalOperator + ", " +
-                    "left: " + left + ", class: " + left.getClass() +
-                    "right: " + right + ", class: " + right.getClass(), e);
+            String msg = String.format("Error on evaluating logical expression. operator: '%s', left: '%s', class: '%s' right: '%s', class: '%s'", logicalOperator, left, left.getClass(), right, right.getClass());
```
[the remark this answers is not in the document]
Comment (bgalek) [reply src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java:52]: sure!

File: src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java [hunk 2]
```diff
@@ -27,7 +27,7 @@ static LogicalOperatorExpressionNode orOperator(OpelNode left, OpelNode right, I
         return left().getValue(context).thenCompose(left -> wrappingExceptionsWithOpelException(left, () -> {
             if (logicalOperator == Operator.OR && Boolean.TRUE.equals(conversion.convert(left, Boolean.class))) {
                 return CompletableFuture.completedFuture(true);
-            } else if (logicalOperator == Operator.AND && Boolean.FALSE.equals(conversion.convert(left, Boolean.class))) {
+            } else if (logicalOperator == Operator.AND && (left == null || Boolean.FALSE.equals(conversion.convert(left, Boolean.class)))) {
```
[the remark this answers is not in the document]
Comment (bgalek) [reply src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java]: moved ;)

File: src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java [hunk 3]
```diff
@@ -40,20 +40,21 @@ static LogicalOperatorExpressionNode orOperator(OpelNode left, OpelNode right, I
...
             return wrappedBody.get();
         } catch (Exception e) {
-            throw new OpelException("Error on evaluating logical expression. " +
-                    "operator: " + logicalOperator + ", " +
-                    "left: " + left + ", class: " + left.getClass() +
-                    "right: " + right + ", class: " + right.getClass(), e);
+            String msg = String.format("Error on evaluating logical expression. " +
+                                       "operator: '%s', left: '%s', class: '%s' right: '%s', class: '%s'",
+                    logicalOperator, left, left.getClass(), right, right.getClass());
+            throw new OpelException(msg, e);
```
Comment (krzysiekbielicki) [inline src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java:57]: What do you think about moving message construction to specific exception?
`OpelLogicalExpressionException(cause: Exception, logicalOperator: Operator, left: Object, right: Object?)`?

**Review by krzysiekbielicki — CHANGES_REQUESTED** [review]

**Review by tjackowiak — APPROVED** [review]

File: src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java (hunk 3 above)
Comment (bgalek) [reply src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java:57]: can do it, but it won't be reusable (both have different messages/args)

Comment (bgalek) [reply src/main/java/pl/allegro/tech/opel/LogicalOperatorExpressionNode.java]: done!

**Pull request merged**